File: sb.js
// WARNING: Using self-bots is strictly against Discord's Terms of Service and can result in your account being permanently banned.
// Use this code at your own risk.

// --- Imports (Using CommonJS 'require' syntax) ---
// Note: We use raw numbers for intents and partials to bypass TypeErrors from the deprecated library's broken exports.
const { 
    Client, 
    MessageAttachment,
    Collection         
} = require('discord.js-selfbot-v13');
const axios = require('axios');
const fs = require('fs');
const path = require('path');

// --- Configuration ---
const PREFIX = '+';
const DELETE_COOLDOWN = 1.0; // Rate limit for self-deletion (1 second delay)
// *** IMPORTANT: This is your updated User Token. Ensure it is valid! ***
const TOKEN = 'MTI1NjIyMDU4NDM4NDEzOTI4NA.GOyhvz.kBk6m1GpG_1EUx8Cwy8wJGv5Z6BAmGFktfcJMA'; 

// --- Client Initialization ---
// FIX APPLIED: Using raw intent number (32767) to bypass broken GatewayIntentBits object.
// FIX APPLIED: Using raw partial number (1) for Partials.Channel to bypass broken Partials object.
const client = new Client({ 
    // 32767 grants all non-privileged intents
    intents: 32767,
    // 1 is the raw numerical value for Partials.Channel
    partials: [1] 
});

// --- State Management ---
const spamTasks = new Map();
const deletedMessages = new Map(); // Stores up to 10 deleted messages per channel
let lastDeleteTime = 0; // Tracks the time of the last successful deletion for rate limiting

// --- Helper Functions ---

/** * Custom message deletion with rate limit handling.
 * This is crucial for self-bots to avoid triggering Discord's internal abuse filters.
 */
async function selfDeleteMessage(message) {
    const currentTime = Date.now() / 1000;
    const timeSinceLastDelete = currentTime - lastDeleteTime;
    
    // Enforce a minimum delay between deletes
    if (timeSinceLastDelete < DELETE_COOLDOWN) {
        // Wait the remaining time
        await new Promise(resolve => setTimeout(resolve, (DELETE_COOLDOWN - timeSinceLastDelete) * 1000));
    }
    
    try {
        await message.delete();
        lastDeleteTime = Date.now() / 1000;
    } catch (error) {
        // Ignore "Unknown Message" error (10008) if message was already deleted
        if (error.code !== 10008) { 
             console.error("Error during message deletion:", error.message);
        }
    }
}

// --- Event Handlers ---

client.on('ready', () => {
    console.log(`Logged in as ${client.user.tag} (ID: ${client.user.id})`);
    console.log('------');
    // Set default presence
    client.user.setStatus('online').catch(console.error);
});

client.on('messageDelete', message => {
    // Check if the message object is complete and not a partial
    if (!message.partial) {
        const channelId = message.channel.id;
        if (!deletedMessages.has(channelId)) {
            deletedMessages.set(channelId, []);
        }
        
        const channelDeletes = deletedMessages.get(channelId);

        // Keep a maximum of 10 deleted messages per channel for the snipe command
        if (channelDeletes.length >= 10) {
            channelDeletes.shift(); 
        }

        channelDeletes.push({
            content: message.content,
            author: message.author,
            timestamp: new Date(),
            // Store URLs of attachments if they exist
            attachments: message.attachments.map(att => att.url)
        });
    }
});

client.on('messageCreate', async message => {
    // CRITICAL: A self-bot MUST only respond to its OWN messages to prevent conflicts and detection
    if (message.author.id !== client.user.id || !message.content.startsWith(PREFIX)) return;

    // Parse the command and arguments
    const args = message.content.slice(PREFIX.length).trim().split(/\s+/);
    const commandName = args.shift().toLowerCase();
    
    // --- Commands implementation starts here ---
    
    if (commandName === 'help') {
        const helpText = `
**MADE BY SLAYERSON NIGGA!**
\`${PREFIX}help\` 

**Messaging Commands:**
• \`${PREFIX}spam <message>\` - Rapidly sends a message (0.01s delay). **EXTREME RISK.**
• \`${PREFIX}stopspam\` - Stops the current spam task
• \`${PREFIX}an <message>\` - Sends an anonymous message (deletes command)
• \`${PREFIX}snipe <amount?>\` - Shows the last deleted message(s) in the channel (max 10 are stored)
• \`${PREFIX}delete <amount>\` - Deletes the bot's own recent messages (max 100)
• \`${PREFIX}cleanconv @user\` - Deletes up to 200 messages sent by the target user OR the bot
• \`${PREFIX}cn @user <amount?>\` - Deletes messages sent by the target user only (max 50, default)

**User Commands:**
• \`${PREFIX}userinfo @user\` - Gets information about a user
• \`${PREFIX}jtbz\` - Sends "Just Too Based!"

**Webhook Commands:**
• \`${PREFIX}checkwebhook <webhookURL>\` - Checks if a webhook URL is valid
• \`${PREFIX}spamwebhook <webhookURL> <webhookName> <message> <amount?>\` - Spams a webhook (max 10)

**Other Commands:**
• \`${PREFIX}account status <online|idle|dnd>\` - Sets the bot's presence status
• \`${PREFIX}serverinfo\` - Gets information about the current server
• \`${PREFIX}joke\` - Tells a random joke
• \`${PREFIX}choose <option1> <option2> ...\` - Chooses a random option (use comma separation for multi-word options)
`;
        // Use a 5-second self-delete on the help message to keep chat clean
        const helpMsg = await message.channel.send(helpText);
        setTimeout(() => selfDeleteMessage(helpMsg), 5000);
    }
    
    // --- Messaging Commands ---

    if (commandName === 'spam') {
        const spamMessage = args.join(' ');
        if (!spamMessage) return message.channel.send('Please provide a message to spam.');

        // Prevent multiple spam tasks in the same channel
        if (spamTasks.has(message.channel.id)) {
            return message.channel.send('A spam task is already running in this channel. Use `+stopspam` first.');
        }
        
        // Delete the command message immediately
        await selfDeleteMessage(message);

        const spamTask = setInterval(() => {
            // WARNING: A 10ms delay is extremely aggressive and WILL cause rate-limiting and potential bans.
            message.channel.send(spamMessage).catch(err => {
                // If a send fails, stop the spam to avoid continuous errors
                console.error("Spam send failed, stopping spam:", err.message);
                clearInterval(spamTasks.get(message.channel.id));
                spamTasks.delete(message.channel.id);
            }); 
        }, 10); 

        spamTasks.set(message.channel.id, spamTask);
    }

    if (commandName === 'stopspam') {
        await selfDeleteMessage(message); 
        if (spamTasks.has(message.channel.id)) {
            clearInterval(spamTasks.get(message.channel.id));
            spamTasks.delete(message.channel.id);
            const tempMsg = await message.channel.send('Spam stopped.');
            setTimeout(() => selfDeleteMessage(tempMsg), 5000);
        } else {
            const tempMsg = await message.channel.send('No spam task is currently running in this channel.');
            setTimeout(() => selfDeleteMessage(tempMsg), 5000);
        }
    }

    if (commandName === 'delete') {
        let amount = parseInt(args[0]) || 10;
        if (amount > 100) amount = 100; // Limit to 100 messages

        // Delete the command message first
        await selfDeleteMessage(message); 

        let deletedCount = 0;
        // Fetch up to 10 more messages than requested to ensure finding enough bot messages
        const messages = await message.channel.messages.fetch({ limit: amount + 10 }); 

        for (const msg of messages.values()) {
            if (deletedCount >= amount) break;
            
            // Only delete messages sent by the self-bot's user ID
            if (msg.author.id === client.user.id) {
                await selfDeleteMessage(msg);
                deletedCount++;
            }
        }
    }
    
    if (commandName === 'an') {
        const anonymousMessage = args.join(' ');
        if (!anonymousMessage) return message.channel.send('Please provide an anonymous message.');
        
        try {
            await selfDeleteMessage(message); // Delete the command
            await new Promise(resolve => setTimeout(resolve, 500)); // Small delay
            await message.channel.send(`Anonymous Message: **${anonymousMessage}**`);
        } catch (e) {
            console.error("Error during anonymous send:", e.message);
        }
    }
    
    if (commandName === 'snipe') {
        await selfDeleteMessage(message);
        let amount = parseInt(args[0]) || 1;
        if (amount < 1) amount = 1;
        
        const channelId = message.channel.id;
        const channelDeletes = deletedMessages.get(channelId);

        if (!channelDeletes || channelDeletes.length === 0) {
            const tempMsg = await message.channel.send("No message to snipe.");
            return setTimeout(() => selfDeleteMessage(tempMsg), 5000);
        }
        
        amount = Math.min(amount, channelDeletes.length);
        // Get the latest 'amount' messages and reverse to display from oldest sniped to newest sniped
        const messagesToSnipe = channelDeletes.slice(-amount).reverse(); 

        for (const [i, msgData] of messagesToSnipe.entries()) {
            const timeDiff = new Date() - msgData.timestamp;
            const secondsAgo = Math.floor(timeDiff / 1000);
            
            const attachments = msgData.attachments || [];
            const attachmentUrls = attachments.length > 0 ? `\n| Attachments: ${attachments.join(', ')}` : "";
            
            const snipeText = `
**SNIPE (${amount - i})**
| Author: ${msgData.author.tag}
| Time: ${secondsAgo} second${secondsAgo !== 1 ? 's' : ''} ago
| Message: ${msgData.content || '*[Content was empty or not captured]**'}
${attachmentUrls}
`;
            const snipeMsg = await message.channel.send(snipeText);
            // Delete snipe results after a delay
            setTimeout(() => selfDeleteMessage(snipeMsg), 10000); 
            await new Promise(resolve => setTimeout(resolve, 300)); // Small delay to avoid rate limit
        }
    }
    
    if (commandName === 'cleanconv' || commandName === 'cn') {
        await selfDeleteMessage(message);
        const user = message.mentions.users.first();
        if (!user) {
             const tempMsg = await message.channel.send("Please mention a user.");
             return setTimeout(() => selfDeleteMessage(tempMsg), 5000);
        }

        const limit = commandName === 'cn' ? parseInt(args[1]) || 50 : 200;
        let messagesToDelete = [];

        // Fetch up to 200 messages (Discord's max fetch limit)
        const messages = await message.channel.messages.fetch({ limit: 200 }); 
        
        for (const msg of messages.values()) {
            const isTarget = commandName === 'cleanconv' 
                ? (msg.author.id === user.id || msg.author.id === client.user.id)
                : (msg.author.id === user.id); 
            
            if (isTarget && msg.id !== message.id) {
                messagesToDelete.push(msg);
            }
        }
        
        messagesToDelete.splice(limit); // Enforce the limit
        let successCount = 0;

        // Deleting messages one by one to respect rate limits
        for (const msg of messagesToDelete) {
            try {
                await selfDeleteMessage(msg);
                successCount++;
                // Add a small delay between deletes to respect rate limit, even with selfDeleteMessage logic
                await new Promise(resolve => setTimeout(resolve, 300)); 
            } catch (e) {
                console.error("Error deleting message during cleanconv:", e.message);
            }
        }
        
        const tempMsg = await message.channel.send(`Successfully deleted ${successCount} message(s) for ${user.tag}.`);
        setTimeout(() => selfDeleteMessage(tempMsg), 5000);
    }


    // --- User Commands ---

    if (commandName === 'userinfo') {
        await selfDeleteMessage(message);
        const user = message.mentions.users.first() || message.author;
        
        const infoText = `
**User Info for ${user.tag}**
**ID:** ${user.id}
**Created:** ${user.createdAt.toLocaleDateString('en-US', { year: 'numeric', month: 'long', day: 'numeric' })}
**Bot:** ${user.bot ? 'Yes' : 'No'}
**Status:** ${user.presence?.status || 'Offline/Unknown'}
`;
        const tempMsg = await message.channel.send(infoText);
        setTimeout(() => selfDeleteMessage(tempMsg), 10000);
    }
    
    if (commandName === 'jtbz') {
        await selfDeleteMessage(message);
        await message.channel.send(" **Just Too Based!** ");
    }


    // --- Webhook Commands ---
    
    if (commandName === 'checkwebhook') {
        const webhookUrl = args[0];
        if (!webhookUrl) return message.channel.send("Please provide a webhook URL.");

        await selfDeleteMessage(message);

        try {
            // GET request to the webhook URL to fetch its data
            const response = await axios.get(webhookUrl);
            if (response.status === 200 && response.data.id) {
                const tempMsg = await message.channel.send(`Webhook Check: **Valid** | Name: ${response.data.name} (ID: ${response.data.id})`);
                setTimeout(() => selfDeleteMessage(tempMsg), 10000);
            } else {
                const tempMsg = await message.channel.send(`Webhook Check: **Invalid** | Status: ${response.status}`);
                setTimeout(() => selfDeleteMessage(tempMsg), 10000);
            }
        } catch (error) {
            const tempMsg = await message.channel.send("Webhook Check: **Error** (Invalid URL format or connection issue)");
            setTimeout(() => selfDeleteMessage(tempMsg), 10000);
        }
    }
    
    if (commandName === 'spamwebhook') {
        const [webhookUrl, webhookName, ...restArgs] = args;
        
        // Try to parse amount from the last argument
        let amount = 5;
        let spamMessage = restArgs.join(' ');
        
        const lastArg = restArgs[restArgs.length - 1];
        if (lastArg && !isNaN(parseInt(lastArg)) && restArgs.length > 0) {
            amount = parseInt(lastArg);
            spamMessage = restArgs.slice(0, restArgs.length - 1).join(' ');
        }
        
        if (!webhookUrl || !webhookName || !spamMessage) {
            await selfDeleteMessage(message);
            const tempMsg = await message.channel.send("Usage: `+spamwebhook <url> <name> <message> <amount?>`");
            return setTimeout(() => selfDeleteMessage(tempMsg), 7000);
        }
        
        await selfDeleteMessage(message); 

        if (amount > 10) {
            amount = 10;
            const tempMsg = await message.channel.send("Limiting spam amount to 10.");
            setTimeout(() => selfDeleteMessage(tempMsg), 5000);
        }
        
        let successCount = 0;
        for (let i = 0; i < amount; i++) {
            try {
                await axios.post(webhookUrl, {
                    content: spamMessage,
                    username: webhookName
                });
                successCount++;
                // IMPORTANT: Webhooks also have rate limits. A 1s delay is safer.
                await new Promise(resolve => setTimeout(resolve, 1000)); 
            } catch (error) {
                console.error(`Webhook spam failed on message ${i+1}: ${error.message}`);
                break; // Stop spamming on the first failure
            }
        }
        
        const tempMsg = await message.channel.send(`Successfully spammed ${successCount} message(s) to webhook.`);
        setTimeout(() => selfDeleteMessage(tempMsg), 5000);
    }


    // --- Other Commands ---

    if (commandName === 'serverinfo') {
        await selfDeleteMessage(message);
        if (!message.guild) return message.channel.send("This command only works in a server.");
        
        const guild = message.guild;
        // Count roles excluding @everyone
        const roles = guild.roles.cache.filter(role => role.name !== '@everyone').size;
        
        const infoText = `
**Server Info for ${guild.name}**
**Owner:** <@${guild.ownerId}>
**ID:** ${guild.id}
**Created:** ${guild.createdAt.toLocaleDateString('en-US', { year: 'numeric', month: 'long', day: 'numeric' })}
**Members:** ${guild.memberCount}
**Roles:** ${roles}
**Channels:** ${guild.channels.cache.size}
**Boost Level:** ${guild.premiumTier || 'None'}
`;
        const tempMsg = await message.channel.send(infoText);
        setTimeout(() => selfDeleteMessage(tempMsg), 10000);
    }
    
    if (commandName === 'joke') {
        await selfDeleteMessage(message);
        const jokes = [
            "Why don't scientists trust atoms? Because they make up everything!",
            "Why did the scarecrow win an award? Because he was outstanding in his field!",
            "What do you call a fake noodle? An impasta!",
            "How does a penguin build its house? Igloos it together!",
            "I told my wife she was drawing her eyebrows too high. She looked surprised."
        ];
        const joke = jokes[Math.floor(Math.random() * jokes.length)];
        const tempMsg = await message.channel.send(joke);
        setTimeout(() => selfDeleteMessage(tempMsg), 10000);
    }

    if (commandName === 'choose') {
        await selfDeleteMessage(message);
        if (args.length === 0) return message.channel.send("Provide options to choose from.");
        
        // Split options by comma and filter out empty strings
        const options = args.join(' ').split(',').map(o => o.trim()).filter(o => o.length > 0);
        
        if (options.length === 0) return message.channel.send("Provide options to choose from.");
        
        const choice = options[Math.floor(Math.random() * options.length)];
        const tempMsg = await message.channel.send(`🎲 I choose: **${choice}**`);
        setTimeout(() => selfDeleteMessage(tempMsg), 10000);
    }

    if (commandName === 'account') {
        await selfDeleteMessage(message);
        const subCommand = args[0]?.toLowerCase();
        
        if (subCommand === 'status') {
            const statusType = args[1]?.toLowerCase();
            const statusMapping = ['online', 'idle', 'dnd', 'invisible'];

            if (!statusType || !statusMapping.includes(statusType)) {
                const tempMsg = await message.channel.send("Invalid status. Use online, idle, or dnd.");
                return setTimeout(() => selfDeleteMessage(tempMsg), 5000);
            }
            
            try {
                await client.user.setStatus(statusType);
                const tempMsg = await message.channel.send(`Status set to **${statusType}**.`);
                setTimeout(() => selfDeleteMessage(tempMsg), 5000);
            } catch (e) {
                const tempMsg = await message.channel.send(`Failed to change status: ${e.message}`);
                setTimeout(() => selfDeleteMessage(tempMsg), 5000);
            }
        } 
        else if (subCommand === 'logout') {
            await message.channel.send("Logging out...").then(() => client.destroy());
        } else {
            const tempMsg = await message.channel.send("Use `+account status <online/idle/dnd>` to change your bot status.");
            setTimeout(() => selfDeleteMessage(tempMsg), 5000);
        }
    }
});


// --- Login ---

// This will attempt to log in using the User Token provided above.
client.login(TOKEN).catch(err => {
    console.error("Failed to log in. Ensure you are using a valid **User Token** and the self-bot library is correctly installed.");
    console.error(err);
});

File: sb.py
#
# Made By Slayerson
#


#
# V6 -All issues fixed 😭😭😭😭
#


import discord
from discord import Intents
from discord.ext import commands
import asyncio
import random
from datetime import datetime
import aiohttp
import json
import time


intents = Intents.default()
intents.message_content = True 
intents.members = True         
intents.presences = True 


bot = commands.Bot(
    command_prefix='+', 
    self_bot=True,   
    help_command=None,
    intents=intents 
)

spam_tasks = {}
an_tasks = {}
deleted_messages = {}
prefixes = {}
last_delete_time = 0
delete_cooldown = 1.0  

@bot.command(name='help')
async def custom_help(ctx):
    help_text = """
**MADE BY SLAYERSON NIGGA!**
`+help` 

**Messaging Commands:**
• `+spam <message>` 
• `+stopspam` 
• `+an <message>` 
• `+stopan` 
• `+snipe <amount?>` 
• `+delete <amount>` 
• `+cleanconv @user` 
• `+cn @user` 

**User Commands:**
• `+userinfo @user` 
• `+jtbz` 

**Webhook Commands:**
• `+checkwebhook <webhookURL>` 
• `+spamwebhook <webhookURL> <webhookName> <message> <amount?>`

**Other Commands:**
• `+account status|profile|logout` 
• `+serverinfo`
• `+joke` - Tell a random joke Look on google for sum Jokes LMFAO
• `+choose <option1> <option2> ...` 
• `+config prefix|settings` 
"""
    await ctx.send(help_text)

@bot.event
async def on_ready():
    
    print(f'Logged in as {bot.user} (ID: {bot.user.id})')
    print('------')

@bot.event
async def on_message_delete(message):
    if message.channel.id not in deleted_messages:
        deleted_messages[message.channel.id] = []
    
    if len(deleted_messages[message.channel.id]) >= 10:
        deleted_messages[message.channel.id].pop(0)
    
    deleted_messages[message.channel.id].append({
        'content': message.content,
        'author': message.author,
        'timestamp': datetime.now(),
        'attachments': [att.url for att in message.attachments]
    })

@bot.command(name='spam')
async def spam_cmd(ctx, *, message):
    if ctx.channel.id in spam_tasks:
        return
    
    async def spam_task():
        try:
            while ctx.channel.id in spam_tasks:
                await ctx.send(message)
                await asyncio.sleep(0.010)
        except Exception as e:
            print(f"Spam error: {e}")
        finally:
            if ctx.channel.id in spam_tasks:
                spam_tasks.pop(ctx.channel.id)
    
    spam_tasks[ctx.channel.id] = asyncio.create_task(spam_task())

@bot.command(name='stopspam')
async def stop_spam(ctx):
    if ctx.channel.id in spam_tasks:
        spam_tasks[ctx.channel.id].cancel()
        spam_tasks.pop(ctx.channel.id)

@bot.command(name='delete')
async def delete_cmd(ctx, amount: int = 10):
    if amount > 100:
        amount = 100
        
    def is_me(m):
        
        return m.author == ctx.author
    
    if isinstance(ctx.channel, discord.DMChannel):
        deleted_count = 0
        async for message in ctx.channel.history(limit=200):
            if deleted_count >= amount:
                break
            if is_me(message) and message.id != ctx.message.id:
                try:
                    await self_delete_message(message)
                    deleted_count += 1
                except:
                    continue
    else:
        
        deleted_count = 0
        async for message in ctx.channel.history(limit=200):
            if deleted_count >= amount:
                break
            if is_me(message) and message.id != ctx.message.id:
                try:
                    await self_delete_message(message)
                    deleted_count += 1
                except:
                    continue


async def self_delete_message(message):
    global last_delete_time
    
    current_time = time.time()
    time_since_last_delete = current_time - last_delete_time
    
    if time_since_last_delete < delete_cooldown:
        await asyncio.sleep(delete_cooldown - time_since_last_delete)
    
    try:
        await message.delete()
        last_delete_time = time.time()
    except discord.HTTPException as e:
        if e.status == 429:
            retry_after = e.retry_after if hasattr(e, 'retry_after') else 2.0
            await asyncio.sleep(retry_after)
            try:
                await message.delete()
                last_delete_time = time.time()
            except:
                pass
        else:
            pass

@bot.group(name='account', invoke_without_command=True)
async def account_group(ctx):
    await ctx.send("Use `+account status <online/idle/dnd>` to change your user status.")

@account_group.command(name='status')
async def set_status(ctx, status_type: str):
    
    status_mapping = {
        'online': 'online',
        'idle': 'idle', 
        'dnd': 'dnd',
        'invisible': 'invisible'
    }
    
    if status_type.lower() not in status_mapping:
        await ctx.send("Invalid status. Use online, idle, dnd, or invisible.")
        return
        
    try:
        
        await bot.change_presence(status=status_type.lower())
        await ctx.send(f"Status set to **{status_type}**.", delete_after=5)
    except Exception as e:
        await ctx.send(f"Failed to change status: {e}", delete_after=5)


@account_group.command(name='logout')
async def logout(ctx):
    await bot.close()

@bot.command(name='an')
async def an_cmd(ctx, *, message):
    try:
        await ctx.message.delete()
        await asyncio.sleep(0.5)
        await ctx.send(f" Anonymous Message: {message}")
    except:
        pass

@bot.command(name='stopan')
async def stop_an(ctx):
    if ctx.channel.id in an_tasks:
        an_tasks[ctx.channel.id].cancel()
        an_tasks.pop(ctx.channel.id)

@bot.command(name='snipe')
async def snipe_cmd(ctx, amount: int = 1):
    if ctx.channel.id not in deleted_messages or not deleted_messages[ctx.channel.id]:
        await ctx.send("No message to snipe.")
        return
    
    amount = min(amount, len(deleted_messages[ctx.channel.id]))
    messages_to_snipe = deleted_messages[ctx.channel.id][-amount:]
    
    for i, msg_data in enumerate(reversed(messages_to_snipe)):
        time_diff = datetime.now() - msg_data['timestamp']
        minutes_ago = int(time_diff.total_seconds() / 60)
        
        attachments = msg_data.get('attachments', [])
        image_install = attachments[0] if attachments else "None"
        
        snipe_text = f"""
**SNIPE**
| Auteur: {msg_data['author']}
| Message: {msg_data['content']}
| Image Install: {image_install}
| Date: {minutes_ago} minute{'s' if minutes_ago != 1 else ''} ago
"""
        await ctx.send(snipe_text)
        await asyncio.sleep(1)

@bot.command(name='cleanconv')
async def clean_conv_cmd(ctx, user: discord.User = None):
    if user is None:
        await ctx.send("Please mention a user.")
        return
        
    def is_conv_message(m):
        
        return m.author == user or m.author == ctx.author
    
    if isinstance(ctx.channel, discord.DMChannel):
        messages_to_delete = []
        async for message in ctx.channel.history(limit=200):
            if is_conv_message(message) and message.id != ctx.message.id:
                messages_to_delete.append(message)
        
        for message in messages_to_delete:
            try:
                await self_delete_message(message)
                await asyncio.sleep(0.3)
            except:
                continue
    else:
        
        messages_to_delete = []
        async for message in ctx.channel.history(limit=100):
            if is_conv_message(message) and message.id != ctx.message.id:
                messages_to_delete.append(message)

        for message in messages_to_delete:
            try:
                await self_delete_message(message)
                await asyncio.sleep(0.3)
            except:
                continue

@bot.command(name='cn')
async def clean_user_messages(ctx, user: discord.User = None, limit: int = 50):
    if user is None:
        await ctx.send("Please mention a user.")
        return
        
    def is_user_message(m):
        return m.author == user
    
    if isinstance(ctx.channel, discord.DMChannel):
        messages_to_delete = []
        async for message in ctx.channel.history(limit=200):
            if is_user_message(message) and message.id != ctx.message.id:
                messages_to_delete.append(message)
        
        for message in messages_to_delete:
            try:
                await self_delete_message(message)
                await asyncio.sleep(0.3)
            except:
                continue
    else:
        
        messages_to_delete = []
        async for message in ctx.channel.history(limit=limit):
            if is_user_message(message) and message.id != ctx.message.id:
                messages_to_delete.append(message)
        
        for message in messages_to_delete:
            try:
                await self_delete_message(message)
                await asyncio.sleep(0.3)
            except:
                continue


@bot.command(name='userinfo')
async def user_info_cmd(ctx, user: discord.User = None):
    user = user or ctx.author
    
    info_text = f"""
**User Info for {user}**
**ID:** {user.id}
**Created:** {user.created_at.strftime('%m/%d/%Y')}
**Bot:** {'Yes' if user.bot else 'No'}
"""
    await ctx.send(info_text)

@bot.command(name='jtbz')
async def jtbz_cmd(ctx):
    await ctx.send(" **Just Too Based!** ")

@bot.command(name='checkwebhook')
async def check_webhook_cmd(ctx, webhook_url: str):
    if isinstance(ctx.channel, discord.DMChannel):
        return
        
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(webhook_url) as response:
                if response.status == 200:
                    data = await response.json()
                    await ctx.send(f"Webhook Check: **Valid** | Name: {data.get('name')}", delete_after=10)
                else:
                    await ctx.send(f"Webhook Check: **Invalid** | Status: {response.status}", delete_after=10)
    except Exception:
        await ctx.send("Webhook Check: **Error** (Invalid URL format or connection issue)", delete_after=10)

@bot.command(name='spamwebhook')
async def spam_webhook_cmd(ctx, webhook_url: str, webhook_name: str, *, message: str, amount: int = 5):
    if isinstance(ctx.channel, discord.DMChannel):
        return
        
    if amount > 10:
        await ctx.send("Limit spam amount to 10.", delete_after=5)
        return
    
    try:
        for i in range(amount):
            async with aiohttp.ClientSession() as session:
                webhook = discord.Webhook.from_url(webhook_url, session=session)
                
                await webhook.send(content=message, username=webhook_name)
                await asyncio.sleep(1)
        await ctx.send(f"Spammed {amount} messages to webhook.", delete_after=5)
    except Exception:
        await ctx.send("Failed to send webhook spam. Check URL and permissions.", delete_after=5)

@bot.command(name='serverinfo')
async def server_info(ctx):
    if isinstance(ctx.channel, discord.DMChannel):
        await ctx.send("This command only works in a server.")
        return
        
    guild = ctx.guild
    roles = [role.name for role in guild.roles if role.name != "@everyone"]
    
    info_text = f"""
**Server Info for {guild.name}**
**Owner:** {guild.owner}
**ID:** {guild.id}
**Created:** {guild.created_at.strftime('%m/%d/%Y')}
**Members:** {guild.member_count}
**Roles:** {len(roles)}
**Channels:** {len(guild.channels)}
"""
    await ctx.send(info_text)

@bot.command(name='joke')
async def tell_joke(ctx):
    jokes = [
        "Why don't scientists trust atoms? Because they make up everything!",
        "Why did the scarecrow win an award? Because he was outstanding in his field!",
        "What do you call a fake noodle? An impasta!",
        "How does a penguin build its house? Igloos it together!",
    ]
    joke = random.choice(jokes)
    await ctx.send(joke)

@bot.command(name='choose')
async def choose_option(ctx, *options):
    if not options:
        await ctx.send("Provide options to choose from.")
        return
        
    choice = random.choice(options)
    await ctx.send(f" I choose: {choice}")

@bot.group(name='config', invoke_without_command=True)
async def config_group(ctx):
    if isinstance(ctx.channel, discord.DMChannel):
        await ctx.send("This command only works in a server.")
        return

@config_group.command(name='prefix')
async def set_prefix(ctx, new_prefix):
    if isinstance(ctx.channel, discord.DMChannel):
        return
        
    prefixes[ctx.guild.id] = new_prefix
    await ctx.send(f"Prefix set to `{new_prefix}` for this server.", delete_after=5)

@bot.event
async def on_command_error(ctx, error):
    if isinstance(error, commands.CommandOnCooldown):
        pass
    elif isinstance(error, discord.HTTPException) and error.status == 429:
        pass
    elif isinstance(error, commands.MissingRequiredArgument):
        pass
    

# DM SLAYERSON FPR HELP


bot.run("add ur shit hete mfs")


# DM SLAYERSON FOR HELP
